cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java

```diff
@@ -283,51 +236,62 @@ public void setStop(final Id.Program id, final String pid, final long endTime, f
   public void setStop(final Id.Program id, final String pid, final long endTime, final ProgramRunStatus runStatus,
                       final BasicThrowable failureCause) {
     Preconditions.checkArgument(runStatus != null, "Run state of program run should be defined");
-    appsTx.get().executeUnchecked(new TransactionExecutor.Function<AppMetadataStore, Void>() {
+    txExecute(transactional, new TxRunnable() {
       @Override
-      public Void apply(AppMetadataStore mds) throws Exception {
-        mds.recordProgramStop(id, pid, endTime, runStatus, failureCause);
-        return null;
-      }
-    }, apps.get());
+      public void run(DatasetContext context) throws Exception {
+        AppMetadataStore metaStore = getAppMetadataStore(context);
+        metaStore.recordProgramStop(id, pid, endTime, runStatus, failureCause);
+
+        // This block has been added so that completed workflow runs can be logged to the workflow dataset
+        if (id.getType() == ProgramType.WORKFLOW && runStatus == ProgramRunStatus.COMPLETED) {
+          RunRecordMeta runRecordMeta = metaStore.getRun(id, pid);
+          if (runRecordMeta == null) {
+            return;
+          }
 
-    // This block has been added so that completed workflow runs can be logged to the workflow dataset
-    if (id.getType() == ProgramType.WORKFLOW && runStatus == ProgramRunStatus.COMPLETED) {
-      Id.Workflow workflow = Id.Workflow.from(id.getApplication(), id.getId());
-      recordCompletedWorkflow(workflow, pid);
-    }
-    // todo: delete old history data
+          recordCompletedWorkflow(metaStore, getWorkflowDataset(context),
```

Comment:
Do I understand this correctly: Old code used multiple transactions, now it all happens in the same tx. Nice.